clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSecurityGroupCachingAgent.groovy

```diff
@@ -53,72 +61,210 @@ class GoogleSecurityGroupCachingAgent extends AbstractGoogleCachingAgent impleme
+    def firewalls = getFirewalls(onDemandSecurityGroupName)
+
+    return firewalls ? firewalls.first() : null
+  }
 
-    CacheResultBuilder cacheResultBuilder = new CacheResultBuilder()
+  private CacheResult buildCacheResult(CacheResultBuilder cacheResultBuilder, List<Firewall> firewalls) {
+    log.info("Describing items in ${agentType}")
 
-    firewallList.collect { Firewall firewall ->
+    firewalls.each { Firewall firewall ->
       def securityGroupKey = Keys.getSecurityGroupKey(firewall.getName(),
                                                       firewall.getName(),
                                                       "global",
                                                       accountName)
 
-      cacheResultBuilder.namespace(SECURITY_GROUPS.ns).keep(securityGroupKey).with {
-        attributes = [firewall: firewall]
+      if (shouldUseOnDemandData(cacheResultBuilder, securityGroupKey)) {
+        moveOnDemandDataToNamespace(cacheResultBuilder, firewall)
+      } else {
+        cacheResultBuilder.namespace(SECURITY_GROUPS.ns).keep(securityGroupKey).with {
+          attributes = [firewall: firewall]
+        }
       }
     }
 
     log.info("Caching ${cacheResultBuilder.namespace(SECURITY_GROUPS.ns).keepSize()} security groups in ${agentType}")
+    log.info "Caching ${cacheResultBuilder.onDemand.toKeep.size()} onDemand entries in ${agentType}"
+    log.info "Evicting ${cacheResultBuilder.onDemand.toEvict.size()} onDemand entries in ${agentType}"
+
+    return cacheResultBuilder.build()
+  }
+
+  static boolean shouldUseOnDemandData(CacheResultBuilder cacheResultBuilder, String securityGroupKey) {
+    CacheData cacheData = cacheResultBuilder.onDemand.toKeep[securityGroupKey]
+
+    return cacheData ? cacheData.attributes.cacheTime >= cacheResultBuilder.startTime : false
+  }
+
+  void moveOnDemandDataToNamespace(CacheResultBuilder cacheResultBuilder, Firewall firewall) {
+    def securityGroupKey = Keys.getSecurityGroupKey(
+      firewall.getName(),
+      firewall.getName(),
+      "global",
+      accountName)
+    Map<String, List<MutableCacheData>> onDemandData = objectMapper.readValue(
+      cacheResultBuilder.onDemand.toKeep[securityGroupKey].attributes.cacheResults as String,
+      new TypeReference<Map<String, List<MutableCacheData>>>() {})
+    onDemandData.each { String namespace, List<MutableCacheData> cacheDatas ->
+      cacheDatas.each { MutableCacheData cacheData ->
+        cacheResultBuilder.namespace(namespace).keep(cacheData.id).with {
+          attributes = cacheData.attributes
+          relationships = cacheData.relationships
+        }
+        cacheResultBuilder.onDemand.toKeep.remove(cacheData.id)
+      }
+    }
+  }
 
-    cacheResultBuilder.build()
+  // TODO(lwander) this was taken from the netflix cluster caching, and should probably be shared between all providers.
+  @Canonical
+  static class MutableCacheData implements CacheData {
```

Comment:
Was just trying to be consistent with the changes you made here: https://github.com/spinnaker/clouddriver/commit/84a0c4a96a4aa9c94f684dc49bad99f7e0547fbf